src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -767,177 +633,105 @@ namespace winrt::Microsoft::Terminal::Control::implementation
-            // the first paint will be ignored!
-            dxEngine->SetWarningCallback(std::bind(&TermControl::_RendererWarning, this, std::placeholders::_1));
+        if (_initializedTerminal)
+        {
+            return false;
+        }
 
-            dxEngine->SetForceFullRepaintRendering(_settings.ForceFullRepaintRendering());
-            dxEngine->SetSoftwareRendering(_settings.SoftwareRendering());
+        const auto panelWidth = SwapChainPanel().ActualWidth();
+        const auto panelHeight = SwapChainPanel().ActualHeight();
+        const auto panelScaleX = SwapChainPanel().CompositionScaleX();
+        const auto panelScaleY = SwapChainPanel().CompositionScaleY();
 
-            // Update DxEngine's AntialiasingMode
-            switch (_settings.AntialiasingMode())
-            {
-            case TextAntialiasingMode::Cleartype:
-                dxEngine->SetAntialiasingMode(D2D1_TEXT_ANTIALIAS_MODE_CLEARTYPE);
-                break;
-            case TextAntialiasingMode::Aliased:
-                dxEngine->SetAntialiasingMode(D2D1_TEXT_ANTIALIAS_MODE_ALIASED);
-                break;
-            case TextAntialiasingMode::Grayscale:
-            default:
-                dxEngine->SetAntialiasingMode(D2D1_TEXT_ANTIALIAS_MODE_GRAYSCALE);
-                break;
-            }
+        const auto windowWidth = panelWidth * panelScaleX;
+        const auto windowHeight = panelHeight * panelScaleY;
 
-            // GH#5098: Inform the engine of the opacity of the default text background.
-            if (_settings.UseAcrylic())
-            {
-                dxEngine->SetDefaultTextBackgroundOpacity(::base::saturated_cast<float>(_settings.TintOpacity()));
-            }
-
-            THROW_IF_FAILED(dxEngine->Enable());
-            _renderEngine = std::move(dxEngine);
+        if (windowWidth == 0 || windowHeight == 0)
+        {
+            return false;
+        }
 
-            _AttachDxgiSwapChainToXaml(_renderEngine->GetSwapChain().Get());
+        // IMPORTANT! Set this callback up sooner than later. If we do it
+        // after Enable, then it'll be possible to paint the frame once
+        // _before_ the warning handler is set up, and then warnings from
+        // the first paint will be ignored!
+        _core->RendererWarning({ get_weak(), &TermControl::_RendererWarning });
 
-            // Tell the DX Engine to notify us when the swap chain changes.
-            // We do this after we initially set the swapchain so as to avoid unnecessary callbacks (and locking problems)
-            _renderEngine->SetCallback(std::bind(&TermControl::RenderEngineSwapChainChanged, this));
+        const auto coreInitialized = _core->InitializeTerminal(panelWidth,
+                                                               panelHeight,
+                                                               panelScaleX,
+                                                               panelScaleY);

```

Comment:
is there an assumption we can make about never scaling x/y differently? might make some param lists shorter.

Reply:
Presumably yea. I'm not sure in practice the scaling has ever actually been different. We assume that it's the same in other places, so I'll de-dupe